kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -179,19 +181,84 @@ private CloseableIterator<ActionWrapper> getNextActionsIter() {
 
                 return combine(dataIter, false /* isFromCheckpoint */, fileVersion);
             } else if (isCheckpointFile(nextFilePath.getName())) {
-                final long fileVersion = checkpointVersion(nextFilePath);
+                CheckpointInstance checkpointInstance =
+                        new CheckpointInstance(nextFilePath.getName());
+                final long fileVersion = checkpointInstance.version;
 
                 // Try to retrieve the remaining checkpoint files (if there are any) and issue
                 // read request for all in one go, so that the {@link ParquetHandler} can do
                 // optimizations like reading multiple files in parallel.
-                CloseableIterator<FileStatus> checkpointFilesIter =
-                        retrieveRemainingCheckpointFiles(nextFile, fileVersion);
-
+                List<FileStatus> checkpointFiles =
+                        retrieveRemainingCheckpointFiles(fileVersion);
+
+                // If the checkpoint file is a UUID-named checkpoint, we may need to read the JSON
+                // actions separately. Otherwise, we can add the checkpoint manifest to the list of
+                // checkpoint files to be read in parallel.
+                CloseableIterator<ColumnarBatch> checkpointJsonActions;
+                if (checkpointInstance.format.equals(CheckpointInstance.CheckpointFormat.V2) &&
+                    nextFilePath.getName().endsWith(".json")) {
+                    checkpointJsonActions = tableClient.getJsonHandler().readJsonFiles(
+                            toCloseableIterator(
+                                    Collections.singletonList(nextFile).iterator()),
+                            readSchema,
+                            Optional.empty());
+                } else {
+                    checkpointFiles.add(nextFile);
+                    checkpointJsonActions = toCloseableIterator(
+                            new ArrayList<ColumnarBatch>().iterator());
+                }
+
+                // Read any potential sidecars for a checkpoint format supporting sidecars and if
+                // the current read is for Add/Remove Files (otherwise, no reason to read sidecars).
+                if (checkpointInstance.format.usesSidecars() &&
+                        LogReplay.isAddRemoveReadSchema(readSchema)) {
+                    // If the checkpoint format may contain sidecars, read the sidecar actions from
+                    // the checkpoint.
+                    final CloseableIterator<ColumnarBatch> manifestIter;
```

Comment:
Should this block be put in a try-with-resources block?

Reply:
if use that, it will be close at the end of the block. We need a mechanism to close them in case there is a failure further down before returning the iterator.